File: 22-generate-parentheses/generate-parentheses.js
/**
 * @param {number} n
 * @return {string[]}
 */
 // 
var generateParenthesis = function(n) {
    let stack = [];
    let result = [];
    function putParenthesis(open, close, n)  {
        if (open === n && close === n) {
            result.push(stack.join(""))
            return
        }
        if (open < n) {
            stack.push("(")
            putParenthesis(open + 1, close, n)
            stack.pop()
        }
        if (close < open) {
            stack.push(")")
            putParenthesis(open , close + 1, n)
            stack.pop()
        }
    }

    putParenthesis(0, 0, n)
    return result
    
};

File: 22-generate-parentheses/generate-parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        # Given n pair of parentheses
        # generate all combinations => recursion

        result = []

        def dfs(s, open, close):
            if open + close == n * 2:
                result.append(s)
            
            if open < n:
                dfs(s + "(", open + 1, close)
            if close < open:
                dfs(s + ")", open, close + 1)
            
            

        dfs("", 0,0)


        return result


        
        